src/tactic/move_add.lean

```diff
@@ -0,0 +1,397 @@
+# `move_add`: a tactic for moving summands
+
+Calling `move_add [a, ← b, c]`, recursively looks inside the goal for expressions involving a sum.
+Whenever it finds one, it moves the summands that unify to `a, b, c`, removing all parentheses.
+
+See the doc-string for `tactic.interactive.move_add` for more information.
+
+##  Implementation notes
+
+This file defines a general `move_op` tactic, intended for reordering terms in an expression
+obtained by repeated applications of a given associative, commutative binary operation.  The
+user decides the final reordering.  Applying `move_op` without specifying the order will simply
+remove all parentheses from the expression.
+The main user-facing tactics are `move_add` and `move_mul`, dealing with addition and
+multiplication, respectively.
+
+In what is below, we talk about `move_add` for definiteness, but everything applies
+to `move_mul` and to the more general `move_op`.
+
+The implementation of `move_add` only moves the terms specified by the user (and rearranges
+parentheses).
+
+Note that the tactic `abel` already implements a very solid heuristic for normalizing terms in an
+additive commutative semigroup and produces expressions in more or less standard form.
+The scope of `move_add` is different: it is designed to make it easy to move individual terms
+around a sum.
+
+##  Future work
+
+* Add support for `neg/div/inv` in additive/multiplicative groups?
+* Customize error messages to mention `move_add/move_mul` instead of `move_op`?
+* Add different operations other than `+` and `*`?  E.g. `∪, ∩, ⊓, ⊔, ...`?
+  Should there be the desire for supporting more operations, it might make sense to extract
+  the `reflexivity <|> simp [add] <|> simp [mul]` block in `sorted_sum` to a separate tactic,
+  including all the lemmas used for the rearrangement to work.
+* Add functionality for moving terms across the two sides of an in/dis/equality.
+  E.g. it might be desirable to have `to_lhs [a]` converting `b + c = a + d` to `- a + b + c = d`.
+* Add a non-recursive version for use in `conv` mode.
+* Revise tests?
+-/
+
+namespace tactic
+
+namespace move_op
+
+/-!
+Throughout this file, `op : pexpr` denotes an arbitrary (binary) operation.  We do not use,
+but implicitly imagine, that this operation is associative, since we extract iterations of
+such operations, with complete disregard of the order in which these iterations arise.
+-/
+
+/-- `list_explicit_args f` returns a list of the explicit arguments of `f`. -/
+meta def list_explicit_args (f : expr) : tactic (list expr) :=
+tactic.fold_explicit_args f [] (λ ll e, return $ ll ++ [e])
+
+/--  `list_head_op op tt e` recurses into the expression `e` looking for first appearances of
+`op` as the head symbol of a subexpression.  Every time it finds one, it isolates it.
+Usually, `op` is a binary, associative operation.  E.g.,
+```lean
+#eval trace $ list_head_op ``((+)) tt `(3 / (2 + 4) + 2 * (0 + 2))
+-- [3 / (2 + 4) + 2 * (0 + 2), 2 + 4, 0 + 2]
```

Comment:
This example isn't a particulary good one, as it doesn't include any terms of the form `a + b + c`